packages/muelu/example/advanced/stratimikos/Stratimikos.cpp

```diff
@@ -182,11 +199,114 @@ int main_(Teuchos::CommandLineProcessor &clp, Xpetra::UnderlyingLib lib, int arg
+  typedef std::complex<double> Scalar;
+#include <MueLu_UseShortNames.hpp>
+
+  using Teuchos::RCP;
+  using Teuchos::rcp;
+
+  //
+  // MPI initialization
+  //
+
+  bool success = false;
+  bool verbose = true;
+  try {
+    RCP< const Teuchos::Comm<int> > comm = Teuchos::DefaultComm<int>::getComm();
+
+    //
+    // Parameters
+    //
+
+    Galeri::Xpetra::Parameters<GlobalOrdinal> matrixParameters(clp, 256); // manage parameters of the test case
+
+    std::string xmlFileName = "stratimikos_ParameterList.xml"; clp.setOption("xml",   &xmlFileName, "read parameters from a file. Otherwise, this example uses by default 'stratimikos_ParameterList.xml'.");
+
+    switch (clp.parse(argc,argv)) {
+      case Teuchos::CommandLineProcessor::PARSE_HELP_PRINTED:        return EXIT_SUCCESS; break;
+      case Teuchos::CommandLineProcessor::PARSE_ERROR:
+      case Teuchos::CommandLineProcessor::PARSE_UNRECOGNIZED_OPTION: return EXIT_FAILURE; break;
+      case Teuchos::CommandLineProcessor::PARSE_SUCCESSFUL:                               break;
+    }
+
+    // Read in parameter list
+    Teuchos::RCP<Teuchos::ParameterList> paramList = Teuchos::getParametersFromXmlFile(xmlFileName);
+
+    //
+    // Construct the problem
+    //
+    RCP<const Map> map = MapFactory::createUniformContigMap(lib, matrixParameters.GetNumGlobalElements(), comm);
+
+    RCP<Galeri::Xpetra::Problem<Map,CrsMatrixWrap,MultiVector> > Pr =
+      Galeri::Xpetra::BuildProblem<SC,LO,GO,Map,CrsMatrixWrap,MultiVector>(matrixParameters.GetMatrixType(), map, matrixParameters.GetParameterList());
+    RCP<CrsMatrixWrap> A = Pr->BuildMatrix();
+
+    RCP<Vector> X = VectorFactory::Build(map);
+    RCP<Vector> B = VectorFactory::Build(map);
+
+    {
+      // we set seed for reproducibility
+      Utilities::SetRandomSeed(*comm);
+      X->randomize();
+      A->apply(*X, *B, Teuchos::NO_TRANS, Teuchos::ScalarTraits<Scalar>::one(), Teuchos::ScalarTraits<Scalar>::zero());
+
+      Teuchos::Array<typename Teuchos::ScalarTraits<Scalar>::magnitudeType> norms(1);
+      B->norm2(norms);
+      B->scale(Teuchos::ScalarTraits<Scalar>::one()/norms[0]);
+      X->putScalar(Teuchos::ScalarTraits<Scalar>::zero());
+    }
+
+    //
+    // Build Thyra linear algebra objects
+    //
+
+    RCP<const Thyra::LinearOpBase<Scalar> > thyraA = Xpetra::ThyraUtils<Scalar,LocalOrdinal,GlobalOrdinal,Node>::toThyra(A->getCrsMatrix());
+
+    RCP<      Thyra::VectorBase<Scalar> >thyraX = Teuchos::rcp_const_cast<Thyra::VectorBase<Scalar> >(Xpetra::ThyraUtils<Scalar,LocalOrdinal,GlobalOrdinal,Node>::toThyraVector(X));
+    RCP<const Thyra::VectorBase<Scalar> >thyraB = Xpetra::ThyraUtils<Scalar,LocalOrdinal,GlobalOrdinal,Node>::toThyraVector(B);
+
+    //
+    // Build Stratimikos solver
+    //
+
+    // Stratimikos::DefaultLinearSolverBuilder linearSolverBuilder;  // This is the Stratimikos main class (= factory of solver factory).
```

Comment:
The partial specialization of `MainWrappers` on `std::complex<double>` doesn't actually do the solve.  Why is this?

Reply:
That's correct. Currently, `Stratimikos::DefaultLinearSolverBuilder` has a hard coded `double`, so I had to comment out that part..

Reply:
Is there a more elegant way to achieve this?

Reply:
@cgcgcg I can't think of one.